### PR TITLE
implement oog call state

### PR DESCRIPTION
**Pull request by DreamWuGit**

### CODE REVIEW

**Comment by DreamWuGit** [comment]
@ed255  or @han0110 may have a look?

File: src/zkevm_specs/evm/execution/oog_call.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+)
+from ..instruction import Instruction, Transition
+from ..table import CallContextFieldTag, AccountFieldTag
+from ..execution_state import ExecutionState
+from ...util import N_BYTES_GAS
+
+
+def oog_call(instruction: Instruction):
+    # retrieve op code associated to oog call error
+    instruction.opcode_lookup(True)
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply src/zkevm_specs/evm/execution/oog_call.py:18]: en, better to check op code is Call. will add it . Thx.

File: tests/evm/test_oog_call.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    return Expected(
+        caller_gas_left=caller_gas_left,
+        callee_gas_left=callee_gas_left + has_value * GAS_STIPEND_CALL_WITH_VALUE,
+        next_memory_size=next_memory_size,
+    )
+
+
+def gen_testing_data():
+    callees = [
+        # CALLEE_WITH_NOTHING,
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply tests/evm/test_oog_call.py:94]: will remove it.Thx!!

File: src/zkevm_specs/evm/execution/oog_call.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply src/zkevm_specs/evm/execution/oog_call.py]: fixed in [ebb7a81](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/287/commits/ebb7a814307be83e9cb425380fab8998f9eead22)

File: tests/evm/test_oog_call.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply tests/evm/test_oog_call.py]: removed in [ebb7a81](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/287/commits/ebb7a814307be83e9cb425380fab8998f9eead22)

File: tests/evm/test_oog_call.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    ["caller_gas_left", "callee_gas_left", "next_memory_size"],
+)
+
+STOP_BYTECODE = Bytecode().stop()
+
+CALLER = Account(address=0xFE, balance=int(1e20))
+CALLEE_WITH_STOP_BYTECODE_AND_BALANCE = Account(address=0xFF, code=STOP_BYTECODE, balance=int(1e18))
+
+
+def expected(callee: Account, caller_ctx: CallContext, stack: Stack, is_warm_access: bool):
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply tests/evm/test_oog_call.py]: this is copied from call test, I realize that it is not required for oog error tests, removed [e21f34e](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/287/commits/e21f34e45fdcbe23f39ddea95773a0a4cb6dd4ab) thanks.

File: src/zkevm_specs/evm/execution/oog_call.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    # verify gas is insufficient
+    gas_not_enough, _ = instruction.compare(instruction.curr.gas_left, gas_cost, N_BYTES_GAS)
+    instruction.constrain_equal(gas_not_enough, FQ(1))
+
+    # current call must be failed.
+    instruction.constrain_equal(
+        instruction.call_context_lookup(CallContextFieldTag.IsSuccess), FQ(0)
+    )
+    is_persistent = instruction.call_context_lookup(CallContextFieldTag.IsPersistent)
+    instruction.constrain_equal(is_persistent, FQ(0))
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/oog_call.py:86]: By https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/815, we can skip this check right?

Comment (DreamWuGit) [reply src/zkevm_specs/evm/execution/oog_call.py:86]: yes, will remove it .

Comment (DreamWuGit) [reply src/zkevm_specs/evm/execution/oog_call.py]: @ChihChengLiang  updated in [8e394ee](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/287/commits/8e394eebb07f32b0fc26e10d4918d835e01575d5) thanks!

**Review by ChihChengLiang — APPROVED** [review]
LGTM after comments are fixed.

File: specs/error_state/ErrorOutOfGasCall.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+below is the total gas cost calculation which from [`Call` Spec](../opcode/F1CALL.md).
+```
+GAS_COST_WARM_ACCESS := 100
+GAS_COST_ACCOUNT_COLD_ACCESS := 2600
+GAS_COST_CALL_EMPTY_ACCOUNT := 25000
+GAS_COST_CALL_WITH_VALUE := 9000
+gas_cost = (
+    GAS_COST_WARM_ACCESS
+    + GAS_COST_WARM_ACCESS if is_warm_access else GAS_COST_ACCOUNT_COLD_ACCESS
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply specs/error_state/ErrorOutOfGasCall.md:20]: right , will update & thanks for catching it ..

File: specs/error_state/ErrorOutOfGasCall.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+## Procedure
+### EVM behavior
+For this gadget, the core is to calculate gas required, there are multiple kinds of gas
+consumes in call: 
+1. memory expansion gas cost
+2. gas cost if new account creates
+3. transfer fee if has value
+4. account access list cost（warm/cold）
+
+below is the total gas cost calculation which from [`Call` Spec](../opcode/F1CALL.md).
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply specs/error_state/ErrorOutOfGasCall.md]: fixed in [30e764a](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/287/commits/30e764a296354280589f9d50afaf78917d1a9003) thanks for catching

File: specs/error_state/ErrorOutOfGasCall.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply specs/error_state/ErrorOutOfGasCall.md]: fixed in [30e764a](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/287/commits/30e764a296354280589f9d50afaf78917d1a9003) thanks !!

**Comment by ChihChengLiang** [comment]
Thanks for the contribution! @DreamWuGit

**Pull request merged**